[JSC] Keep builtin UnlinkedFunctionExecutable

Constellation opened the pull request:
#### 6165ff108d4cb0a08d1989951f6af78efa5644e4
<pre>
[JSC] Keep builtin UnlinkedFunctionExecutable
<a href="https://bugs.webkit.org/show_bug.cgi?id=271324">https://bugs.webkit.org/show_bug.cgi?id=271324</a>
<a href="https://rdar.apple.com/125101756">rdar://125101756</a>

Reviewed by Justin Michaud and Keith Miller.

We found that they go away quickly when opening a new iframe etc., and as a result, we may parse
and generate code again and again for very popular builtin functions in some cases. Let&apos;s keep
them in a similar manner to CodeCache. And clear when CodeCache clearing gets requested.

* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::visitAggregateImpl):
(JSC::BuiltinExecutables::clear):
(JSC::BuiltinExecutables::finalizeUnconditionally): Deleted.
* Source/JavaScriptCore/builtins/BuiltinExecutables.h:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::finalizeUnconditionalFinalizers):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::deleteAllCode):
(JSC::VM::visitAggregateImpl):

Canonical link: <a href="https://commits.webkit.org/276432@main">https://commits.webkit.org/276432@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1026bd4b206f37587ecaed6b52e9eb1efb79a472

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44632 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23711 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47088 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47285 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40637 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27749 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21128 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45208 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20774 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38417 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/17764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18221 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39559 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2679 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/37807 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40849 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39842 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48946 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/44077 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19585 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16144 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20916 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/42419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51259 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6161 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20584 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10397 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 reviewed (approved):
r=me.

justinmichaud reviewed (approved):
r=me

hyjorc1 commented on Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
```diff
@@ -265,14 +265,22 @@ UnlinkedFunctionExecutable* BuiltinExecutables::createExecutable(VM& vm, const S
-        if (unlinkedExecutable && !m_vm.heap.isMarked(unlinkedExecutable))
-            unlinkedExecutable = nullptr;
+        if (unlinkedExecutable)
+            visitor.appendUnbarriered(unlinkedExecutable);
```
What are `visitAggregate` functions used for? And what is `appendUnbarriered` trying to achineve?

Constellation committed:
[JSC] Keep builtin UnlinkedFunctionExecutable
https://bugs.webkit.org/show_bug.cgi?id=271324
rdar://125101756

Reviewed by Justin Michaud and Keith Miller.

We found that they go away quickly when opening a new iframe etc., and as a result, we may parse
and generate code again and again for very popular builtin functions in some cases. Let's keep
them in a similar manner to CodeCache. And clear when CodeCache clearing gets requested.

* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::visitAggregateImpl):
(JSC::BuiltinExecutables::clear):
(JSC::BuiltinExecutables::finalizeUnconditionally): Deleted.
* Source/JavaScriptCore/builtins/BuiltinExecutables.h:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::finalizeUnconditionalFinalizers):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::deleteAllCode):
(JSC::VM::visitAggregateImpl):

Canonical link: https://commits.webkit.org/276432@main
